### PR TITLE
chore: added merge task logic and tests

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔/✖                                                                        |
| New feature     | ✔/✖                                                                        |
| Breaking change | ✔/✖                                                                        |
| Deprecations    | ✔/✖                                                                        |
| Documentation   | ✔/✖                                                                        |
| Tests added     | ✔/✖                                                                        |
| Chore            | ✔/✖                                                                       |

Related issues: #XXX , #XXX ...
Closes #XXX ...

Further  information:

### CODE REVIEW

File: tests/unit/layers/models/tasker.spec.ts [hunk 1]
```diff
@@ -64,6 +67,32 @@ describe('Tasker', () => {
     initConfig();
   });
 
+  describe('createIngestionTask', () => {
```
Comment (asafMasa) [inline tests/unit/layers/models/tasker.spec.ts:70]: update remark to the actual ingestion / job task types

File: src/serviceClients/jobManagerClient.ts [hunk 2]
```diff
@@ -104,7 +108,7 @@ export class JobManagerClient extends HttpClient {
       productType: data.metadata.productType,
       tasks: taskParams?.map((params) => {
         return {
-          type: taskType,
+          type: jobType === JobType.UPDATE ? TaskType.MERGE : TaskType.DISCRETE_TILING,
```
Comment (asafMasa) [inline src/serviceClients/jobManagerClient.ts:111]: type - should come from config

File: src/serviceClients/jobManagerClient.ts [hunk 3]
```diff
@@ -118,7 +122,18 @@ export class JobManagerClient extends HttpClient {
...
+      };
+    });
+    await this.post(createTasksUrl, req);
+  }
+
+  public async createMergeTasks(jobId: string, taskParams: IMergeTaskParams[]): Promise<void> {
+    const createTasksUrl = `/jobs/${jobId}/tasks`;
+    const req = taskParams.map((params) => {
+      return {
+        type: TaskType.MERGE,
```
Comment (asafMasa) [inline src/serviceClients/jobManagerClient.ts:136]: type - should come from config

File: src/layers/models/tasker.ts [hunk 4]
```diff
@@ -6,14 +6,61 @@ import { ITaskParameters } from '../interfaces';
...
   private readonly bboxSizeTiles: number;
-  public constructor(@inject(Services.CONFIG) private readonly config: IConfig) {
+  private readonly tasksBatchSize: number;
+
+  public constructor(@inject(Services.CONFIG) private readonly config: IConfig, private readonly db: JobManagerClient) {
     this.bboxSizeTiles = config.get<number>('bboxSizeTiles');
+    this.tasksBatchSize = config.get<number>('tasksBatchSize');
   }
 
+  public async createIngestionTask(
```
Comment (asafMasa) [inline src/layers/models/tasker.ts:22]: method name is unclear, i would suggest 'createTilesSplittingTask'

File: src/layers/models/tasker.ts [hunk 5]
```diff
@@ -6,14 +6,61 @@ import { ITaskParameters } from '../interfaces';
...
+    jobType: JobType
+  ): Promise<void> {
+    const taskParams = this.generateTasksParameters(data, layerRelativePath, layerZoomRanges);
+    let taskBatch: ITaskParameters[] = [];
+    let jobId: string | undefined = undefined;
+    for (const task of taskParams) {
+      taskBatch.push(task);
+      if (taskBatch.length === this.tasksBatchSize) {
+        if (jobId === undefined) {
+          jobId = await this.db.createLayerJob(data, layerRelativePath, jobType, taskBatch);
```
Comment (asafMasa) [inline src/layers/models/tasker.ts:35]: send configured task type to method

File: src/layers/models/tasker.ts [hunk 6]
```diff
@@ -6,14 +6,61 @@ import { ITaskParameters } from '../interfaces';
...
+        } else {
+          // eslint-disable-next-line no-useless-catch
+          try {
+            await this.db.createTasks(jobId, taskBatch);
```
Comment (asafMasa) [inline src/layers/models/tasker.ts:39]: send configured task type to method

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/common/interfaces.ts [hunk 7]
```diff
@@ -40,8 +39,11 @@ export interface IMergeOverlaps {
   intersection: GeoJSON;
 }
 
+export interface IMergeSources {
+  type: string;
+  path: string;
+}
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/common/interfaces.ts:45]: fixed

File: package.json [hunk 8]
```diff
@@ -49,8 +49,10 @@
     "@map-colonies/mc-utils": "^1.4.3",
     "@map-colonies/storage-explorer-middleware": "^1.2.3",
     "@turf/turf": "^6.5.0",
+    "ansi-regex": "^6.0.1",
     "axios": "^0.26.1",
     "axios-retry": "^3.2.4",
+    "better-sqlite3": "^7.5.1",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply package.json:55]: removed

File: package.json [hunk 9]
```diff
@@ -49,8 +49,10 @@
     "@map-colonies/mc-utils": "^1.4.3",
     "@map-colonies/storage-explorer-middleware": "^1.2.3",
     "@turf/turf": "^6.5.0",
+    "ansi-regex": "^6.0.1",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply package.json:52]: removed

File: src/common/interfaces.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/common/interfaces.ts:45]: fixed

File: package.json [hunk 10]
```diff
@@ -68,6 +70,7 @@
     "@commitlint/config-conventional": "^10.0.0",
     "@map-colonies/eslint-config": "^2.2.0",
     "@redocly/openapi-cli": "^1.0.0-beta.25",
+    "@types/better-sqlite3": "^7.5.0",
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply package.json:73]: removed

File: src/common/interfaces.ts [hunk 11]
```diff
@@ -23,7 +23,6 @@ export interface IFindResponseRecord extends IRasterCatalogUpsertRequestBody {
 }
 
 export type FindRecordResponse = IFindResponseRecord[];
-
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/common/interfaces.ts:26]: fixed

File: src/serviceClients/jobManagerClient.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/serviceClients/jobManagerClient.ts:136]: fixed

File: src/layers/models/tasker.ts (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/tasker.ts:22]: fixed

File: src/layers/models/tasker.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/tasker.ts:35]: fixed

File: src/layers/models/tasker.ts (hunk 6 above)
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/tasker.ts:39]: fixed

File: src/layers/models/fileValidator.ts [hunk 12]
```diff
@@ -22,4 +22,14 @@ export class FileValidator {
     const allValid = (await Promise.all(filePromises)).every((value) => value);
     return allValid;
   }
+
+  public async validateGpkgFiles(files: string[]): Promise<boolean> {
+    const gpkgExt = '.gpkg';
+    const filePromises = files.map((file) => {
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/fileValidator.ts:28]: fixed

File: src/layers/models/layersManager.ts [hunk 13]
```diff
@@ -51,56 +48,54 @@ export class LayersManager {
...
     const resourceId = data.metadata.productId as string;
     const version = data.metadata.productVersion as string;
     const productType = data.metadata.productType as ProductType;
 
-    await this.validateNotRunning(resourceId, version, productType);
+    await this.validateNotRunning(resourceId, version, productType, jobType);
     await this.validateNotExistsInCatalog(resourceId, version, productType);
-    await this.validateNotExistsInMapServer(resourceId, productType);
     await this.validateFiles(data);
+    await this.validateNotExistsInMapServer(resourceId, productType);
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/layersManager.ts:98]: fixed - moved only to "New" ingestion job logic

File: src/layers/models/layersManager.ts [hunk 14]
```diff
@@ -51,56 +48,54 @@ export class LayersManager {
     this.logger.log('info', `creating job and tasks for layer ${data.metadata.productId as string}`);
     const layerRelativePath = `${data.metadata.productId as string}/${data.metadata.productType as string}`;
     const layerZoomRanges = this.zoomLevelCalculator.createLayerZoomRanges(data.metadata.maxResolutionDeg as number);
-    await this.createTasks(data, layerRelativePath, layerZoomRanges);
+    await this.createTasks(data, layerRelativePath, layerZoomRanges, jobType);
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/layersManager.ts:51]: fixed

File: src/layers/models/tasker.ts [hunk 15]
```diff
@@ -6,14 +6,61 @@ import { ITaskParameters } from '../interfaces';
 import { ITaskZoomRange } from '../../tasks/interfaces';
 import { Services } from '../../common/constants';
 import { IConfig } from '../../common/interfaces';
+import { JobType, OperationStatus } from '../../common/enums';
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/tasker.ts:9]: fixed

File: src/serviceClients/jobManagerClient.ts [hunk 16]
```diff
@@ -77,7 +77,6 @@ interface IGetJobResponse {
 }
 
 const jobType = config.get<string>('jobType');
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/serviceClients/jobManagerClient.ts:79]: fixed

File: src/layers/models/tasker.ts [hunk 17]
```diff
@@ -6,14 +6,61 @@ import { ITaskParameters } from '../interfaces';
...
 
 @singleton()
 export class Tasker {
   private readonly bboxSizeTiles: number;
-  public constructor(@inject(Services.CONFIG) private readonly config: IConfig) {
+  private readonly tasksBatchSize: number;
+
+  public constructor(@inject(Services.CONFIG) private readonly config: IConfig, private readonly db: JobManagerClient) {
     this.bboxSizeTiles = config.get<number>('bboxSizeTiles');
+    this.tasksBatchSize = config.get<number>('tasksBatchSize');
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/tasker.ts]: fixed

File: src/common/enums.ts [hunk 18]
```diff
@@ -12,3 +12,19 @@ export enum MapServerCacheType {
...
 }
+
+export enum JobType {
+  DISCRETE_TILING = 'Discrete-Tiling',
+  UPDATE = 'Update',
+}
+
+export enum TaskType {
+  DISCRETE_TILING = 'Discrete-Tiling',
+  MERGE = 'Merge',
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/common/enums.ts]: fixed

File: src/layers/models/layersManager.ts [hunk 19]
```diff
@@ -51,56 +48,54 @@ export class LayersManager {
...
+    }
+  }
+
+  private async validateRunConditions(data: IngestionParams, jobType: JobType): Promise<void> {
     const resourceId = data.metadata.productId as string;
     const version = data.metadata.productVersion as string;
     const productType = data.metadata.productType as ProductType;
 
-    await this.validateNotRunning(resourceId, version, productType);
+    await this.validateNotRunning(resourceId, version, productType, jobType);
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/layersManager.ts]: fixed

File: src/layers/models/layersManager.ts [hunk 20]
```diff
@@ -51,56 +48,54 @@ export class LayersManager {
...
+        );
       }
     }
+    return false;
   }
 
-  private async validateRunConditions(data: IngestionParams): Promise<void> {
+  private async createTasks(data: IngestionParams, layerRelativePath: string, layerZoomRanges: ITaskZoomRange[], jobType: JobType): Promise<void> {
+    if (jobType === JobType.UPDATE) {
+      const allValidGpkgs = await this.fileValidator.validateGpkgFiles(data.fileNames);
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/layers/models/layersManager.ts]: fixed

File: src/common/enums.ts [hunk 21]
```diff
@@ -12,3 +12,19 @@ export enum MapServerCacheType {
   FS = 'FS',
   S3 = 'S3',
 }
+
+export enum JobType {
+  DISCRETE_TILING = 'Discrete-Tiling',
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply src/common/enums.ts]: fixed

File: tests/unit/layers/models/tasker.spec.ts [hunk 22]
```diff
@@ -1,9 +1,12 @@
 import { IngestionParams, LayerMetadata, ProductType, RecordType } from '@map-colonies/mc-model-types';
+import { JobType } from '../../../../src/common/enums';
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply tests/unit/layers/models/tasker.spec.ts:2]: fixed

File: tests/unit/layers/models/layersModel.spec.ts [hunk 23]
```diff
@@ -65,155 +67,138 @@ const testData: IngestionParams = {
...
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
     clearMockConfig();
     initMockConfig();
+    checkForUpdateSpy = jest.spyOn(LayersManager.prototype, 'checkForUpdate');
   });
 
   describe('createLayer', () => {
-    it('saves metadata before queueing tasks', async function () {
+    it('saves metadata before queueing ingestion tasks', async function () {
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply tests/unit/layers/models/layersModel.spec.ts:78]: fixed

File: tests/unit/merger/mergeTasker.spec.ts [hunk 24]
```diff
@@ -197,61 +266,101 @@ describe('mergeTasker', () => {
...
-        destPath: 'test/dest',
-        maxZoom: 1,
-      };
-
-      const taskGen = mergeTasker.createBatchedTasks(params);
-      const tasks: IMergeTaskParams[] = [];
-      for (const task of taskGen) {
-        tasks.push(task);
-      }
+    // it('same footprint', () => {
```
[the remark this answers is not in the document]
Comment (CL-SHLOMIKONCHA) [reply tests/unit/merger/mergeTasker.spec.ts]: fixed

File: tests/unit/layers/models/tasker.spec.ts (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply tests/unit/layers/models/tasker.spec.ts:70]: fixed

File: src/layers/models/splitTilesTasker.ts [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+    this.splitTilesTaskType = config.get<string>('ingestionNewTaskType');
+  }
+
+  public async createSplitTilesTasks(
+    data: IngestionParams,
+    layerRelativePath: string,
+    layerZoomRanges: ITaskZoomRange[],
+    jobType: string
+  ): Promise<void> {
+    this.setDefaultValues(data);
```
Comment (asafMasa) [inline src/layers/models/splitTilesTasker.ts:33]: move back to layers manager

File: src/merge/mergeTilesTasker.ts [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+              })
+            ),
+          };
+        }
+      }
+    }
+  }
+
+  public async createMergeTilesTasks(data: IngestionParams, layerRelativePath: string): Promise<void> {
+    const sourceDir = this.config.get<string>('LayerSourceDir');
```
Comment (asafMasa) [inline src/merge/mergeTilesTasker.ts:112]: move the config values reading to constructor

File: src/merge/mergeTilesTasker.ts [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+      maxZoom: maxZoom,
+    };
+    const mergeTasksParams = this.createBatchedTasks(params);
+    let mergeTaskBatch: IMergeTaskParams[] = [];
+    let jobId: string | undefined = undefined;
+    for (const mergeTask of mergeTasksParams) {
+      mergeTaskBatch.push(mergeTask);
+      if (mergeTaskBatch.length === this.mergeTaskBatchSize) {
+        if (jobId === undefined) {
+          jobId = await this.db.createLayerJob(data, layerRelativePath, jobType, this.mergeTilesTaskType, mergeTaskBatch);
```
Comment (asafMasa) [inline src/merge/mergeTilesTasker.ts:136]: add task type here

File: src/merge/mergeTilesTasker.ts [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+        } else {
+          try {
+            await this.db.createMergeTasks(jobId, mergeTaskBatch);
```
Comment (asafMasa) [inline src/merge/mergeTilesTasker.ts:139]: add task type here

File: src/merge/mergeTilesTasker.ts [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+            await this.db.updateJobStatus(jobId, OperationStatus.FAILED);
+            throw err;
+          }
+        }
+        mergeTaskBatch = [];
+      }
+    }
+    if (mergeTaskBatch.length !== 0) {
+      if (jobId === undefined) {
+        jobId = await this.db.createLayerJob(data, layerRelativePath, jobType, this.mergeTilesTaskType, mergeTaskBatch);
```
Comment (asafMasa) [inline src/merge/mergeTilesTasker.ts:150]: add task type here

File: src/merge/mergeTilesTasker.ts [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+      } else {
+        // eslint-disable-next-line no-useless-catch
+        try {
+          await this.db.createMergeTasks(jobId, mergeTaskBatch);
```
Comment (asafMasa) [inline src/merge/mergeTilesTasker.ts:154]: add task type here

File: src/serviceClients/jobManagerClient.ts [hunk 31]
```diff
@@ -76,8 +76,9 @@ interface IGetJobResponse {
   abortedTasks: number;
 }
 
-const jobType = config.get<string>('jobType');
-const taskType = config.get<string>('taskType');
+const ingestionNewTaskType = config.get<string>('ingestionNewTaskType');
```
Comment (asafMasa) [inline src/serviceClients/jobManagerClient.ts:79]: ?

File: src/serviceClients/jobManagerClient.ts [hunk 32]
```diff
@@ -118,7 +125,18 @@ export class JobManagerClient extends HttpClient {
     const createTasksUrl = `/jobs/${jobId}/tasks`;
     const req = taskParams.map((params) => {
       return {
-        type: taskType,
+        type: ingestionNewTaskType,
```
Comment (asafMasa) [inline src/serviceClients/jobManagerClient.ts:128]: get it from method args

File: src/serviceClients/jobManagerClient.ts [hunk 33]
```diff
@@ -118,7 +125,18 @@ export class JobManagerClient extends HttpClient {
...
+        parameters: params,
+      };
+    });
+    await this.post(createTasksUrl, req);
+  }
+
+  public async createMergeTasks(jobId: string, taskParams: IMergeTaskParams[]): Promise<void> {
```
Comment (asafMasa) [inline src/serviceClients/jobManagerClient.ts:135]: duplicated method - remove

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: tests/integration/layers/layers.spec.ts [hunk 34]
```diff
@@ -70,7 +70,7 @@ describe('layers', function () {
   beforeEach(function () {
     console.warn = jest.fn();
     setValue('tiling.zoomGroups', '0,1,2,3,4,5,6,7,8,9,10');
-    setValue('tasksBatchSize', 2);
+    setValue('ingestionNewTiles.tasksBatchSize', 2);
```
[the remark this answers is not in the document]
Comment (asafMasa) [reply tests/integration/layers/layers.spec.ts:73]: fix LGTM warnings
split tasker
image

merge tasker
image

File: src/layers/models/layersManager.ts [hunk 35]
```diff
@@ -1,106 +1,122 @@
...
-      data.metadata.layerPolygonParts = layerMetadataToPolygonParts(data.metadata);
+    await this.validateFiles(data);
+    await this.validateJobNotRunning(resourceId, productType);
+
+    this.logger.log('info', `creating ${jobType} job and ${taskType} tasks for layer ${data.metadata.productId as string} type: ${productType}`);
+
+    if (jobType === JobType.NEW) {
+      const layerZoomRanges = this.zoomLevelCalculator.createLayerZoomRanges(data.metadata.maxResolutionDeg as number);
+
+      await this.validateNotExistsInCatalog(resourceId, version, productType);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:60]: move validations first, no need to call "createLayerZoomRanges" if the operation isn't valid

File: src/layers/models/layersManager.ts [hunk 36]
```diff
@@ -1,106 +1,122 @@
...
+      await this.validateNotExistsInMapServer(resourceId, productType);
+      this.setDefaultValues(data);
+
+      await this.splitTilesTasker.createSplitTilesTasks(data, layerRelativePath, layerZoomRanges, jobType, taskType);
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    } else if (jobType === JobType.UPDATE) {
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:66]: if you already added the else if, add else {
.. throw unsupported jobType 
}

**Review by asafMasa — CHANGES_REQUESTED** [review]
looking better, but i still have more changes...
mind some of my old remarks weren't resolved

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Pull request merged**